IQSS/dataverse-frontend
# 95 - Invoke js dataverse use cases in the dataset page

## What this PR does / why we need it:
This PR adds the use cases of the js-dataverse module to the Dataset page. This means that we're now able to show real datasets in the frontend.

This PR also improves the containerized development environment

## Which issue(s) this PR closes:

- Closes #95
- Closes #106 
- Closes #77

## Special notes for your reviewer:
### Changes dev-env
- Get js-dataverse from published npm registry
- Add withData flag when running dev-env

### Changes datasets UI
- Point e2e tests to the development environment port
- Refactor Datasets use cases to filter datasets by persistentId, version or privateUrlToken
- Fix Dataset Labels to show correctly the Dataset labels depending on the version state
- Fix metadata fields values to accept all compound fields formats
- Add a Dataset Builder class to create the dataset labels depending on the version
- Create a JSDatasetMapper class to map js-daverse format to UI format
- Replace mocks from Datasets use cases with the real js-daverse implementations
- Set 'metadata' as the default tab when opening a dataset page until 'files' tab is developed
- Refactor DatasetCitation component to show the citation html directly instead of composing the citation
- Replace the .tgz file of the js-dataverse module by the npm registry version
- Divide e2e into e2e and integration folders
- Fix old e2e tests and add new ones

## Suggestions on how to test this:
### Step 1: Get a personal token to use the js-dataverse registry
1. create a copy of `.npmrc.example` file without the .example suffix
2. go to https://github.com
3. Click your profile image
4. Go to settings
5. In the lateral bar click Developer Settings
6. In the lateral bar click Personal Access Tokens and select Tokens(classic)
7. Click Generate new token (classic)
8. Add a name to your new token
9. Select read:packages check
10. Click Generate token button
11. Copy your token
12. Substitute your new token in the .npmrc file
14. To check if it works run `npm i`

### Step 2: Run the application locally
1. Run `npm i`
2. `cd packages/design-system && npm run build`
3. cd back to the root.
4. Check that you have a  `.env` file such as the .env.example, with the `VITE_DATAVERSE_BACKEND_URL=http://localhost:8000` variable
5. `npm start`
6. Go to <http://localhost:5173/spa> to check that the app runs locally. Stop it with Ctrl-c afterwards.

### Step 3: Run the development environment
1. Install jq: Download the binary for your platform from http://stedolan.github.io/jq/ and make sure it is in your $PATH (/usr/bin/jq is fine) and executable with `sudo chmod +x /usr/bin/jq`. On Mac, you can install it with `brew install jq` if you use homebrew: http://brew.sh/ .
2.  `cd dev-env`
3. `./run-env.sh 9588-datasets-api-extension`

If it gives a timeout before the containers are ready, run again the above command

4. To check the environment go to <http://localhost:8000> and check your local dataverse installation

### Step 4: Test Dataset View mode
1. Go to <http://localhost:8000>
4. Login as admin using username: dataverseAdmin and password: admin1
5. Create a new Dataset
6. From the dataset view mode copy the search parameters from the url. Ex.: `?persistentId=doi:10.5072/FK2/LHGRHP&version=DRAFT`
7. Go to <http://localhost:8000/spa/datasets> and paste the search parameters. Ex.: <http://localhost:8000/spa/datasets?persistentId=doi:10.5072/FK2/LHGRHP&version=DRAFT>
8. You are seeing now the same Dataset in the SPA, check that all the data is displayed correctly

### Step 4: Test an specific dataset version

1. To test a specific dataset version in the version search parameter add your exisiting version. Ex.: <http://localhost:8000/spa/datasets?persistentId=doi:10.5072/FK2/LHGRHP&version=1.0>

### Step 5: Test a dataset using a privateUrlToken
1. Create a new Dataset in localhost:8000
2. With the Dataset in draft mode click the button Edit Dataset > Private Url > Create Private Url
3. From the url generated copy only the token search parameter value
4. Now to see the dataset with the private url in the SPA go to the datasets url and paste the privateUrlToken search parameter <http://localhost:8000/spa/datasets?privateUrlToken=dfa0efbc-2a54-49e0-b16e-1bafb4c4d770>

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:

## Is there a release notes update needed for this change?:
Display a dataset by persistentId or privateUrlToken in the SPA

Improve the development environment of the SPA

## Additional documentation:
